src/hotspot/share/cds/archiveHeapWriter.cpp

```diff
@@ -186,47 +182,71 @@ void ArchiveHeapWriter::ensure_buffer_space(size_t min_bytes) {
-    oopDesc::set_mark(mem, markWord::prototype());
-    oopDesc::release_set_klass(mem, k);
-  }
-  {
-    // This is copied from ObjArrayAllocator::initialize
-    arrayOopDesc::set_length(mem, length);
+void ArchiveHeapWriter::root_segment_at_put(objArrayOop segment, int index, oop root) {
+  // Do not use arrayOop->obj_at_put(i, o) as arrayOop is outside the real heap!
+  if (UseCompressedOops) {
+    *segment->obj_at_addr<narrowOop>(index) = CompressedOops::encode(root);
+  } else {
+    *segment->obj_at_addr<oop>(index) = root;
   }
+}
 
-  objArrayOop arrayOop = objArrayOop(cast_to_oop(mem));
-  for (int i = 0; i < length; i++) {
-    // Do not use arrayOop->obj_at_put(i, o) as arrayOop is outside of the real heap!
-    oop o = roots->at(i);
-    if (UseCompressedOops) {
-      * arrayOop->obj_at_addr<narrowOop>(i) = CompressedOops::encode(o);
-    } else {
-      * arrayOop->obj_at_addr<oop>(i) = o;
+void ArchiveHeapWriter::copy_roots_to_buffer(GrowableArrayCHeap<oop, mtClassShared>* roots) {
+  // Depending on the number of classes we are archiving, a single roots array may be
+  // larger than MIN_GC_REGION_ALIGNMENT. Roots are allocated first in the buffer, which
+  // allows us to chop the large array into a series of "segments". Current layout
+  // starts with zero or more segments exactly fitting MIN_GC_REGION_ALIGNMENT, and end
+  // with a single segment that may be smaller than MIN_GC_REGION_ALIGNMENT.
+  // This is simple and efficient. We do not need filler objects anywhere between the segments,
+  // or immediately after the last segment. This allows starting the object dump immediately
+  // after the roots.
+
+  assert((_buffer_used % MIN_GC_REGION_ALIGNMENT) == 0,
+         "Pre-condition: Roots start at aligned boundary: " SIZE_FORMAT, _buffer_used);
+
+  int max_elem_count = ((MIN_GC_REGION_ALIGNMENT - arrayOopDesc::header_size_in_bytes()) / heapOopSize);
+  assert(objArrayOopDesc::object_size(max_elem_count)*HeapWordSize == MIN_GC_REGION_ALIGNMENT,
+         "Should match exactly");
+
+  HeapRootSegments segments(_buffer_used,
+                            roots->length(),
+                            MIN_GC_REGION_ALIGNMENT,
+                            max_elem_count);
+
+  for (size_t seg_idx = 0; seg_idx < segments.count(); seg_idx++) {
+    int size_elems = segments.size_in_elems(seg_idx);
+    size_t size_bytes = segments.size_in_bytes(seg_idx);
+
+    size_t oop_offset = _buffer_used;
+    _buffer_used = oop_offset + size_bytes;
+    ensure_buffer_space(_buffer_used);
+
+    assert((oop_offset % MIN_GC_REGION_ALIGNMENT) == 0,
+           "Roots segment " SIZE_FORMAT " start is not aligned: " SIZE_FORMAT,
+           segments.count(), oop_offset);
+
+    int seg_start = segments.roots_offset(seg_idx);
+    objArrayOop seg_oop = manifest_root_segment(oop_offset, size_elems);
+    for (int i = 0; i < size_elems; i++) {
+      root_segment_at_put(seg_oop, i, roots->at(seg_start + i));
```

Comment:
`HeapRootSegments::roots_offset()` is used only here, so I think it's better to avoid having an extra function that you have to understand by reading both of it definition and it usage here. Maybe this can be simplified with:

```
int root_index = 0;
for (size_t seg_idx = 0; seg_idx < segments.count(); seg_idx++) {
   ...
    for (int i = 0; i < size_elems; i++) {
      root_segment_at_put(seg_oop, i, roots->at(root_index++));
 ```

Reply:
Right, that is cleaner. Would do this in new commit.